File: .claude/skills/gemini-imagegen/scripts/edit.py
#!/usr/bin/env python3
# /// script
# dependencies = [
#   "google-genai",
#   "pillow",
# ]
# ///
"""Edit an existing image with a text prompt using Gemini API.

Usage:
    op run -- uv run scripts/edit.py input.jpg "Add a sunset to the sky" -o edited.jpg
    op run -- uv run scripts/edit.py photo.png "Make it look like a watercolor painting" --size 2K
"""

import argparse
import os
import sys

from PIL import Image
from google import genai
from google.genai import types


def edit_image(
    input_path: str,
    prompt: str,
    output_path: str = "edited.jpg",
    image_size: str = "1K",
    model: str = "gemini-3-pro-image-preview",
) -> str:
    """Edit an existing image with a text prompt.

    Args:
        input_path: Path to the input image
        prompt: Text description of the edits to make
        output_path: Path to save the edited image (use .jpg extension)
        image_size: Resolution (1K, 2K, or 4K)
        model: Gemini model to use

    Returns:
        Path to the saved image
    """
    client = genai.Client(api_key=os.environ["GEMINI_API_KEY"])

    # Load the input image
    input_image = Image.open(input_path)

    config = types.GenerateContentConfig(
        response_modalities=["TEXT", "IMAGE"],
        image_config=types.ImageConfig(
            image_size=image_size,
        ),
    )

    response = client.models.generate_content(
        model=model,
        contents=[prompt, input_image],
        config=config,
    )

    for part in response.parts:
        if part.text:
            print(part.text)
        elif part.inline_data:
            image = part.as_image()
            image.save(output_path)
            print(f"Edited image saved to: {output_path}")
            return output_path

    raise RuntimeError("No image was generated")


def main():
    parser = argparse.ArgumentParser(
        description="Edit an image with a text prompt using Gemini API"
    )
    parser.add_argument("input", help="Path to the input image")
    parser.add_argument("prompt", help="Text description of the edits to make")
    parser.add_argument(
        "-o", "--output",
        default="edited.jpg",
        help="Output file path (default: edited.jpg)"
    )
    parser.add_argument(
        "--size",
        default="1K",
        choices=["1K", "2K", "4K"],
        help="Image resolution (default: 1K)"
    )
    parser.add_argument(
        "--model",
        default="gemini-3-pro-image-preview",
        help="Gemini model to use"
    )

    args = parser.parse_args()

    if not os.path.exists(args.input):
        print(f"Error: Input file not found: {args.input}", file=sys.stderr)
        sys.exit(1)

    try:
        edit_image(
            input_path=args.input,
            prompt=args.prompt,
            output_path=args.output,
            image_size=args.size,
            model=args.model,
        )
    except Exception as e:
        print(f"Error: {e}", file=sys.stderr)
        sys.exit(1)


if __name__ == "__main__":
    main()

File: .agents/skills/imagegen/scripts/edit.ts
#!/usr/bin/env bun

import { parseArgs } from "node:util";

import {
  DEFAULT_MODEL,
  IMAGE_SIZES,
  RESPONSE_MODALITIES,
  createClient,
  ensureImageSize,
  inlineImagePart,
  saveFirstImageFromResponse,
} from "./common";

function printUsage(error?: string): void {
  const stream = error ? process.stderr : process.stdout;
  if (error) {
    console.error(`Error: ${error}`);
  }
  stream.write(
    `Usage: skillx imagegen edit <input-image> "prompt" [-o edited.jpg] [--size ${IMAGE_SIZES.join(
      "|",
    )}] [--model ${DEFAULT_MODEL}]\n`,
  );
}

type EditCliArgs = {
  inputPath: string;
  prompt: string;
  outputPath: string;
  imageSize: string;
  model: string;
};

type ParseResult =
  | {
      type: "ok";
      value: EditCliArgs;
    }
  | {
      type: "exit";
      code: number;
    };

function parseCliArgs(argv: string[]): ParseResult {
  const parsed = parseArgs({
    args: argv,
    allowPositionals: true,
    options: {
      output: { type: "string", short: "o", default: "edited.jpg" },
      size: { type: "string", default: "1K" },
      model: { type: "string", default: DEFAULT_MODEL },
      help: { type: "boolean", short: "h", default: false },
    },
  });

  if (parsed.values.help) {
    printUsage();
    return { type: "exit", code: 0 };
  }

  if (parsed.positionals.length < 2) {
    printUsage("Missing required arguments: <input-image> and/or prompt");
    return { type: "exit", code: 2 };
  }

  if (parsed.positionals.length > 2) {
    printUsage("Unexpected extra positional arguments");
    return { type: "exit", code: 2 };
  }

  ensureImageSize(parsed.values.size);

  return {
    type: "ok",
    value: {
      inputPath: parsed.positionals[0],
      prompt: parsed.positionals[1],
      outputPath: parsed.values.output,
      imageSize: parsed.values.size,
      model: parsed.values.model,
    },
  };
}

export async function runEdit(argv: string[]): Promise<number> {
  try {
    const parsed = parseCliArgs(argv);
    if (parsed.type === "exit") {
      return parsed.code;
    }
    const { inputPath, prompt, outputPath, imageSize, model } = parsed.value;

    const client = createClient();
    const response = await client.models.generateContent({
      model,
      contents: [prompt, inlineImagePart(inputPath)],
      config: {
        responseModalities: RESPONSE_MODALITIES,
        imageConfig: {
          imageSize,
        },
      },
    });

    saveFirstImageFromResponse(response, outputPath, "Edited image saved to");
    return 0;
  } catch (error) {
    const message = error instanceof Error ? error.message : String(error);
    console.error(`Error: ${message}`);
    return 1;
  }
}

if (import.meta.main) {
  process.exit(await runEdit(process.argv.slice(2)));
}
